### PR TITLE
Componetize Nebula and Glean

### PR DESCRIPTION
**Pull request by mkestler-mozilla**
I need someone to do the macOS/iOS/android additions to the project files.  I took a stab at Linux but need to verify.  Glean and Nebula are static lib projects now.

### CODE REVIEW

**Review by bakulf — COMMENTED** [review]
What is the PDB? Is it needed?

File: crashreporter/.gitignore [hunk 1]
```diff
@@ -0,0 +1,73 @@
+# This file is used to ignore files which are generated
```
Comment (bakulf) [inline crashreporter/.gitignore:1]: Do we need it? Can we use the same .gitignore at the root of the repo?

File: crashreporter/MozVPNCrashReporter.pro [hunk 2]
```diff
@@ -0,0 +1,82 @@
+QT       += core gui
```
Comment (bakulf) [inline crashreporter/MozVPNCrashReporter.pro:1]: rename the file `crashreporter.pro`

Comment (bakulf) [inline crashreporter/MozVPNCrashReporter.pro:1]: Add:
```
# This Source Code Form is subject to the terms of the Mozilla Public
# License, v. 2.0. If a copy of the MPL was not distributed with this
# file, You can obtain one at http://mozilla.org/MPL/2.0/.
```

File: crashreporter/MozVPNCrashReporter.pro [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+        QMAKE_CXXFLAGS += /Z7 /ZI /FdMozVPNCarashReporter.PDB /DEBUG
+        QMAKE_LFLAGS_WINDOWS += /DEBUG
+    }
+
+
+}
+
+
+# Default rules for deployment.
+qnx: target.path = /tmp/$${TARGET}/bin
```
Comment (bakulf) [inline crashreporter/MozVPNCrashReporter.pro:51]: qnx? :) I guess we don't want to support this platform yet.

File: crashreporter/MozVPNCrashReporter.pro [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+else: unix:!android: target.path = /opt/$${TARGET}/bin
```
Comment (bakulf) [inline crashreporter/MozVPNCrashReporter.pro:52]: For linux:

```
isEmpty(USRPATH) {
    USRPATH=/usr
}
target.path = $${USRPATH}/bin
```

File: crashreporter/MozVPNCrashReporter.pro [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+!isEmpty(target.path): INSTALLS += target
+
+#translations
+RESOURCES += $$PWD/../translations/servers.qrc
```
Comment (bakulf) [inline crashreporter/MozVPNCrashReporter.pro:56]: I don't think we need this. This is the list of translations for the countries/cities

File: crashreporter/main.cpp [hunk 6]
```diff
@@ -0,0 +1,23 @@
+#include "mainwindow.h"
```
Comment (bakulf) [inline crashreporter/main.cpp:1]:
```
/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
```

File: crashreporter/mainwindow.cpp [hunk 7]
```diff
@@ -0,0 +1,11 @@
+#include "mainwindow.h"
```
Comment (bakulf) [inline crashreporter/mainwindow.cpp:1]: header here and in all the other files.

File: glean.vcxproj [hunk 8]
```diff
@@ -0,0 +1,173 @@
+<?xml version="1.0" encoding="utf-8"?>
```
Comment (bakulf) [inline glean.vcxproj:1]: Do we need this file? Tell me more. I though it was generated by the compilation scripts.

File: glean.vcxproj.filters [hunk 9]
```diff
@@ -0,0 +1,89 @@
+<?xml version="1.0" encoding="utf-8"?>
```
Comment (bakulf) [inline glean.vcxproj.filters:1]: Same here.

File: glean/glean.h [hunk 10]
```diff
@@ -0,0 +1,16 @@
+#ifndef GLEAN_H
+#define GLEAN_H
+
+#include <QQmlEngine>
+
+class Glean
+{
+private:
+    Glean(){};
```
Comment (bakulf) [inline glean/glean.h:9]: Glean() = default;

File: crashreporter/mainwindow.h [hunk 11]
```diff
@@ -0,0 +1,14 @@
+#ifndef MAINWINDOW_H
+#define MAINWINDOW_H
+
+#include <QMainWindow>
+
+class MainWindow : public QMainWindow
```
Comment (bakulf) [inline crashreporter/mainwindow.h:6]: final

File: crashreporter/mainwindow.h [hunk 12]
```diff
@@ -0,0 +1,14 @@
...
+{
+    Q_OBJECT
+
+public:
+    MainWindow(QWidget *parent = nullptr);
```
Comment (bakulf) [inline crashreporter/mainwindow.h:11]: explicit

File: nebula/nebula.h [hunk 13]
```diff
@@ -0,0 +1,27 @@
+#ifndef NEBULA_H
+#define NEBULA_H
+
+#include <QQmlEngine>
+
+class Nebula
+{
+private:
+    Nebula(){};
```
Comment (bakulf) [inline nebula/nebula.h:9]: = default;

File: nebula/nebula.h [hunk 14]
```diff
@@ -0,0 +1,27 @@
+#ifndef NEBULA_H
+#define NEBULA_H
+
+#include <QQmlEngine>
+
+class Nebula
```
Comment (bakulf) [inline nebula/nebula.h:6]: final

File: nebula/ui/resources/settings-white.svg [hunk 15]
```diff
@@ -0,0 +1,3 @@
+<svg width="24" height="24" viewBox="0 0 24 24" fill="none" xmlns="http://www.w3.org/2000/svg">
```
Comment (bakulf) [inline nebula/ui/resources/settings-white.svg:1]: is this a new file?

File: src/commands/commandui.cpp [hunk 16]
```diff
@@ -124,10 +127,9 @@ int CommandUI::run(QStringList& tokens) {
     // This object _must_ live longer than MozillaVPN to avoid shutdown crashes.
     QmlEngineHolder engineHolder;
     QQmlApplicationEngine* engine = QmlEngineHolder::instance()->engine();
-    engine->addImportPath("qrc:///components");
-    engine->addImportPath("qrc:///glean");
-    engine->addImportPath("qrc:///themes");
-    engine->addImportPath("qrc:///compat");
+
+    Glean::Initialize(engine);
+    Nebula::Initialize(engine);
```
Comment (bakulf) [inline src/commands/commandui.cpp:132]: nice!

File: src/commands/commandui.cpp [hunk 17]
```diff
@@ -407,7 +409,8 @@ int CommandUI::run(QStringList& tokens) {
         engine, &QQmlApplicationEngine::objectCreated, qApp,
         [url](QObject* obj, const QUrl& objUrl) {
           if (!obj && url == objUrl) {
-            QGuiApplication::exit(-1);
+            logger.error() << "Failed to load " << objUrl.toString();
+            // QGuiApplication::exit(-1);
```
Comment (bakulf) [inline src/commands/commandui.cpp:413]: why this?

File: src/src.pro [hunk 18]
```diff
@@ -314,31 +313,26 @@ unix {
     HEADERS += signalhandler.h
 }
 
-RESOURCES += ui/components.qrc
-RESOURCES += ui/license.qrc
+RESOURCES += inspector/inspector.qrc
```
Comment (bakulf) [inline src/src.pro:316]: remove this :) it's gone!

**Review by bakulf — CHANGES_REQUESTED** [review]
Can we land this without crashpad?

File: nebula/ui/resources/settings-white.svg (hunk 15 above)
Comment (mkestler-mozilla) [reply nebula/ui/resources/settings-white.svg:1]: I couldn't track changes in resources easily.  git mv got confused as I moved things back and forth.  It is needed by nebula.  It came from the old resource qrc.

File: src/commands/commandui.cpp (hunk 17 above)
Comment (mkestler-mozilla) [reply src/commands/commandui.cpp]: Accidental commit.  Sorry.  Was useful for QML debugging not to exit there.

**Comment by mkestler-mozilla** [comment]
> 
> 
> Can we land this without crashpad?

Yes.  This was inadvertent.  That is just a dummy project to make sure linking works.

**Comment by mbirghan** [comment]
Both iOS and macOS work now.
@bakulf Would you take a look at my changes to make sure that they are ok?

**Review by bakulf — APPROVED** [review]

**Pull request merged**